apps/api/src/indexing/events.repository.ts

```diff
@@ -0,0 +1,136 @@
+ *     using the natural chain key (block_number, tx_hash, log_index).
+ *   - `save()` inserts the normalized raw event payload (append-only),
+ *     relying on a UNIQUE (tx_hash, log_index) in the schema to enforce
+ *     at-most-once writes even under races.
+ *   - `lastBlockNumber()` returns the resume point for the indexer,
+ *     falling back to configured `startBlock` if the table is empty.
+ *
+ * Relationships:
+ * - Called by `EventProcessor.processEvent()` *after* stake mutations succeed.
+ * - Read by `IndexingService.loop()` to set `fromBlock = last + 1`.
+ *
+ * Conventions:
+ * - Hashes and addresses must be lower-cased before insert (done here).
+ * - Timestamps are UTC (`timestamp` without time zone).
+ */
+export class EventsRepository {
+  readonly #db: typeof db;
+
+  constructor(database: typeof db = db) {
+    this.#db = database;
+  }
+
+  /**
+   * Check if a chain log has already been recorded in `indexing_events`.
+   *
+   * Idempotency key:
+   * - (block_number, transaction_hash, log_index)
+   *
+   * Returns:
+   * - true  → already present (caller should skip processing)
+   * - false → new event (caller may proceed)
+   *
+   * Notes:
+   * - `transactionHash` is normalized to lowercase to match storage convention.
+   * - Fast due to index on (tx_hash, log_index) and/or block_number.
+   */
+  async isEventPresent(
+    blockNumber: bigint,
+    transactionHash: string,
+    logIndex: number,
+  ): Promise<boolean> {
+    const txHash = TxHashCoder.encode(transactionHash);
+    const rows = await this.#db
+      .select({ id: indexingEvents.blockNumber })
+      .from(indexingEvents)
+      .where(
+        and(
+          eq(indexingEvents.blockNumber, blockNumber),
+          eq(indexingEvents.transactionHash, txHash),
+          eq(indexingEvents.logIndex, logIndex),
+        ),
+      )
+      .limit(1);
+    return rows.length > 0;
+  }
+
+  /**
+   * Append a normalized raw event into `indexing_events`.
+   *
+   * Behavior:
+   * - Inserts the event with all chain coordinates + untouched `raw` payload.
+   * - Uses `onConflictDoNothing()` so duplicate inserts are safely ignored
+   *   (race-proof w.r.t. concurrent indexer workers).
+   *
+   * Returns:
+   * - true  → row inserted (first time we see this log)
+   * - false → insert was skipped by the unique constraint (already present)
+   *
+   * Caller contract:
+   * - Should only be called **after** domain mutations (stakes/stake_changes)
+   *   have succeeded; this preserves “apply state, then record intake” semantics.
+   * - All hashes are lower-cased here; upstream code can pass mixed case.
+   */
+  async append(event: EventData): Promise<boolean> {
+    const txHash = TxHashCoder.encode(event.transactionHash);
+    const blockHash = BlockHashCoder.encode(event.blockHash);
+    const rows = await this.#db
+      .insert(indexingEvents)
+      .values({
+        id: crypto.randomUUID(),
```

Comment:
random observations: we use `import { v4 as uuidv4 } from "uuid";` and leverage `uuidv4()` pretty much everywhere else in our code. but, i suppose `crypto.randomUUID()` might be the more "correct" option since its faster/native to nodejs.

i suppose it doesnt _really_ matter, unless we want to be consistent on how the `id` is generated here / other queries.

Reply:
Two options here:
- better use v7 to reflect creation timeline, but it is a bit slower;
- crypto.randomUUID is the fastest, and it does the same v4.